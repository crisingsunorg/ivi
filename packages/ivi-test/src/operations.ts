import { OpNode, NodeFlags, Op, OpArray, ElementData, ContextData, Key } from "ivi";

export function visitOpNodes(
  op: Op,
  parent: Op,
  key: any,
  context: {},
  visitor: (vnode: Op, parent: Op, key: any, context: {}) => boolean,
): boolean {
  if (visitor(op, parent, key, context) === true) {
    return true;
  }

  if (typeof op === "object" && op !== null) {
    if (op instanceof Array) {
      for (let i = 0; i < op.length; i++) {
        if (visitOpNodes(op[i], op, void 0, context, visitor)) {
          return true;
        }
      }
    } else {
      const flags = op.type.flags;
      if ((flags & NodeFlags.Element) !== 0) {
        return visitOpNodes(op.data.children, op, void 0, context, visitor);
      } else if ((flags & NodeFlags.TrackByKey) !== 0) {
        const children = (op as OpNode<Key<any, Op>[]>).data;
        for (let i = 0; i < children.length; i++) {
          const k = children[i];
          if (visitOpNodes(k.v, op, k.k, context, visitor)) {
            return true;
          }
        }
      } else if ((flags & NodeFlags.Context) !== 0) {
        return visitOpNodes(op.data.children, op, void 0, { ...context, ...op.data.data }, visitor);
      } else if ((flags & (NodeFlags.Events | NodeFlags.Ref)) !== 0) {
        return visitOpNodes(op.data.children, op, void 0, context, visitor);
      }
    }
  }

  return false;
}

export function isOpText(op: Op): op is string | number {
  return (typeof op === "string" || typeof op === "number");
}

export function isOpObject(op: Op): op is OpNode | OpArray {
  return (typeof op === "object" && op !== null);
}

export function isOpFragment(op: Op): op is OpArray {
  return isOpObject(op) && op instanceof Array;
}

export function isOpElement(op: Op): op is OpNode<ElementData> {
  return isOpObject(op) && !isOpFragment(op) && (op.type.flags & NodeFlags.Element) !== 0;
}

export function isOpComponent<T = any>(op: Op): op is OpNode<T> {
  return isOpObject(op) && !isOpFragment(op) && (op.type.flags & NodeFlags.Component) !== 0;
}

export function isOpContext(op: Op): op is OpNode<ContextData> {
  return isOpObject(op) && !isOpFragment(op) && (op.type.flags & NodeFlags.Component) !== 0;
}
