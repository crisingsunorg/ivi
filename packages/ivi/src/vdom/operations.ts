import { Box } from "../core/box";
import { EventHandler } from "../events/event_handler";
import { NodeFlags } from "./node_flags";
import { StateNode } from "./state";
import { ElementProtoDescriptor } from "./element_proto";
import { ComponentDescriptor } from "./component";

/**
 * Operation type.
 */
export interface OpType {
  /**
   * See {@link NodeFlags} for details.
   */
  readonly flags: NodeFlags;
  /**
   * Operation type descriptor.
   */
  readonly descriptor: ComponentDescriptor | ElementProtoDescriptor | string | null;
}

/**
 * createOpType creates {@link OpType} instances.
 *
 * @param flags See {@link NodeFlags} for details.
 * @param descriptor Operation type descriptor.
 * @returns {@link OpType} instance.
 */
export const createOpType = (
  flags: NodeFlags,
  descriptor: ComponentDescriptor | ElementProtoDescriptor | string | null,
): OpType => ({ flags, descriptor });

/**
 * OpType for Events nodes.
 */
export const EVENTS = createOpType(NodeFlags.Events, null);

/**
 * OpType for Ref nodes.
 */
export const REF = createOpType(NodeFlags.Ref, null);

/**
 * OpType for Context nodes.
 */
export const CONTEXT = createOpType(NodeFlags.Context, null);

/**
 * OpType for TrackByKey nodes.
 */
export const TRACK_BY_KEY = createOpType(NodeFlags.TrackByKey | NodeFlags.MultipleChildren, null);

/**
 * Operation node.
 */
export interface OpNode<T = any> {
  /**
   * Operation type.
   */
  readonly type: OpType;
  /**
   * Operation data.
   */
  readonly data: T;
}

/**
 * createOpNode creates an {@link OpNode} instance.
 *
 * @param type Operation type.
 * @param data Operation data.
 * @returns {@link OpNode} instance.
 */
export const createOpNode = <T>(type: OpType, data: T): OpNode<T> => ({ type, data });

/**
 * Operation data for element operations.
 */
export interface ElementData<T = any> {
  /**
   * Element class name.
   */
  readonly className: string | undefined;
  /**
   * Element attributes.
   */
  readonly attrs: T | undefined;
  /**
   * Children operations.
   */
  readonly children: OpChildren;
}

/**
 * Children operations supported by element nodes.
 */
export type OpChildren = string | number | OpNode | RecursiveOpChildrenArray | null;

/**
 * Recursive children array operations supported by element nodes.
 */
export interface RecursiveOpChildrenArray extends Array<OpChildren> { }

/**
 * Generic operation data for simple operations.
 */
export interface OpData<T = any> {
  /**
   * Generic data.
   */
  readonly data: T;
  /**
   * Child operation.
   */
  readonly child: OpNode | string | number | null;
}

/**
 * Operation data for Events operations.
 */
export type EventsData = OpData<EventHandler | Array<EventHandler | null> | null>;

/**
 * Operation data for Ref operations.
 */
export type RefData = OpData<Box<StateNode>>;

/**
 * Operation data for Context operations.
 */
export type ContextData = OpData<{}>;

/**
 * Operation factory for event handlers.
 *
 * @example
 *
 *     render(
 *       Events(onClick(() => { console.log("clicked"); }),
 *         ChildComponent(),
 *       ),
 *       DOMContainer,
 *     );
 *
 * @param data Event handlers.
 * @param child Child operation.
 * @returns Event handler OpNode.
 */
export const Events = DEBUG ?
  (
    data: EventHandler | Array<EventHandler | null> | null,
    child: OpNode | string | number | null,
  ): OpNode<EventsData> => {
    if (child !== null && typeof child === "object" && child.type === TRACK_BY_KEY) {
      throw new Error(`Invalid child OpNode, Events can't have TrackByKey as a child`);
    }
    return createOpNode(EVENTS, { data, child });
  } :
  (
    data: EventHandler | Array<EventHandler | null> | null,
    child: OpNode | string | number | null,
  ): OpNode<EventsData> => (
      createOpNode(EVENTS, { data, child })
    );

/**
 * Operation factory for ref nodes.
 *
 * @example
 *
 *     const _ref = box();
 *
 *     render(
 *       Ref(_ref,
 *         div(),
 *       ),
 *       DOMContainer,
 *     );
 *
 *     getDOMNode(_ref);
 *
 * @param data Boxed value.
 * @param child Child operation.
 * @returns Ref OpNode.
 */
export const Ref = DEBUG ?
  (
    data: Box<StateNode>,
    child: OpNode | string | number | null,
  ): OpNode<RefData> => {
    if (child !== null && typeof child === "object" && child.type === TRACK_BY_KEY) {
      throw new Error(`Invalid child OpNode, Ref can't have TrackByKey as a child`);
    }
    return createOpNode(REF, { data, child });
  } :
  (
    data: Box<StateNode>,
    child: OpNode | string | number | null,
  ): OpNode<RefData> => createOpNode(REF, { data, child });

/**
 * Operation factory for context nodes.
 *
 * @example
 *
 *     render(
 *       Context({ key: 123 },
 *         ChildComponent(),
 *       ),
 *       DOMContainer,
 *     );
 *
 * @param data Context object.
 * @param child Child operation.
 * @returns Context OpNode.
 */
export const Context = DEBUG ?
  (data: {}, child: OpNode | string | number | null): OpNode<ContextData> => {
    if (child !== null && typeof child === "object" && child.type === TRACK_BY_KEY) {
      throw new Error(`Invalid child OpNode, Context can't have TrackByKey as a child`);
    }
    return createOpNode(CONTEXT, { data, child });
  } :
  (data: {}, child: OpNode | string | number | null): OpNode<ContextData> => createOpNode(CONTEXT, { data, child });

/**
 * Key is an object that is used by TrackByKey operations to track operations.
 */
export interface Key<K, V> {
  /**
   * Key.
   */
  readonly k: K;
  /**
   * Value.
   */
  readonly v: V;
}

/**
 * key creates a {@link Key} instance.
 *
 * @param k Key.
 * @param v Value.
 * @returns {@link Key} instance.
 */
export const key = <K, V>(k: K, v: V): Key<K, V> => ({ k, v });

/**
 * Operation factory for track by key nodes.
 *
 * @example
 *
 *     const items = [1, 2, 3];
 *     render(
 *       TrackByKey(items.map((item) => key(item, div(_, _, item))),
 *       DOMContainer,
 *     );
 *
 * @param items Keyed operations.
 * @returns Track by key OpNode.
 */
export const TrackByKey = DEBUG ?
  <T>(items: Key<T, OpNode | number | string>[]) => {
    if (DEBUG) {
      const keys = new Set<T>();
      for (let i = 0; i < items.length; i++) {
        const { k, v } = items[i];
        if (keys.has(k)) {
          throw new Error(`Invalid key, found duplicated key: ${k}`);
        }
        keys.add(k);
        if (typeof v === "object" && v.type === TRACK_BY_KEY) {
          throw new Error(`Invalid child OpNode, TrackByKey can't have TrackByKey children`);
        }
      }
    }
    return createOpNode(TRACK_BY_KEY, items);
  } :
  <T>(items: Key<T, OpNode | number | string>[]) => createOpNode(TRACK_BY_KEY, items);
