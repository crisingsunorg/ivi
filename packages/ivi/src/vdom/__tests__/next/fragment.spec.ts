import {
  useComputedValue, useResetJSDOM, useResetModules, useDOMElement, useDOMOpsCounters, useIVI, useHTML, useTest,
} from "ivi-jest";
import { Op } from "../../operations";

useResetJSDOM();
useResetModules();
const c = useDOMElement();
const domOps = useDOMOpsCounters();
const ivi = useIVI();
const h = useHTML();
const t = useTest();
const Component = useComputedValue(() => ivi.component<Op>(() => (ops) => ops));
const r = (op: Op) => t.render(op, c());

describe("fragment", () => {
  describe("mount", () => {
    test("empty", () => {
      r([]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("single child", () => {
      r([h.h1()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("two children", () => {
      r([h.h1(), h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("null between two children", () => {
      r([h.h1(), null, h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    describe("components", () => {
      test("first null", () => {
        r([Component()(null), Component()(h.h1()), Component()(h.h2())]);
        expect(c()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("last null", () => {
        r([Component()(h.h1()), Component()(h.h2()), Component()(null)]);
        expect(c()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });

      test("null between two children", () => {
        r([Component()(h.h1()), Component()(null), Component()(h.h2())]);
        expect(c()).toMatchSnapshot();
        expect(domOps()).toMatchSnapshot();
      });
    });
  });

  describe("update", () => {
    test("null to two children", () => {
      r(null);
      r([h.h1(), h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("null to fragment with a hole", () => {
      r(null);
      r([null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("one child to two children", () => {
      r([h.h1()]);
      r([h.h1(), h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("two children to one child", () => {
      r([h.h1(), h.h2()]);
      r([h.h1()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 1", () => {
      r([null, null]);
      r([h.h1(), null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 2", () => {
      r([null, null]);
      r([null, h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 3", () => {
      r([null, null]);
      r([h.h1(), h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 4", () => {
      r([h.h1(), null]);
      r([null, null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 5", () => {
      r([null, h.h2()]);
      r([null, null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 6", () => {
      r([h.h1(), h.h2()]);
      r([null, null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 7", () => {
      r([h.h1(), null]);
      r([null, h.h2()]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });

    test("hole 8", () => {
      r([null, h.h2()]);
      r([h.h1(), null]);
      expect(c()).toMatchSnapshot();
      expect(domOps()).toMatchSnapshot();
    });
  });
});
