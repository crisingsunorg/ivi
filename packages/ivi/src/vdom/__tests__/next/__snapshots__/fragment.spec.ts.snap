// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragment mount components first null 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components first null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount components last null 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components last null 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount components null between two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount components null between two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount empty 1`] = `<div />`;

exports[`fragment mount empty 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 0,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount null between two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount null between two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount single child 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment mount single child 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment mount two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment mount two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 1 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update hole 1 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 2 1`] = `
<div>
  <h2 />
</div>
`;

exports[`fragment update hole 2 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 3 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update hole 3 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 4 1`] = `<div />`;

exports[`fragment update hole 4 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 5 1`] = `<div />`;

exports[`fragment update hole 5 2`] = `
Object {
  "appendChild": 0,
  "createElement": 1,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 1,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 6 1`] = `<div />`;

exports[`fragment update hole 6 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 2,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 7 1`] = `
<div>
  <h2 />
</div>
`;

exports[`fragment update hole 7 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update hole 8 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update hole 8 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 1,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update null to fragment with a hole 1`] = `<div />`;

exports[`fragment update null to fragment with a hole 2`] = `
Object {
  "appendChild": 0,
  "createElement": 0,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 0,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update null to two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update null to two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 2,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 2,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 0,
}
`;

exports[`fragment update one child to two children 1`] = `
<div>
  <h1 />
  <h2 />
</div>
`;

exports[`fragment update one child to two children 2`] = `
Object {
  "appendChild": 0,
  "createElement": 3,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 1,
}
`;

exports[`fragment update two children to one child 1`] = `
<div>
  <h1 />
</div>
`;

exports[`fragment update two children to one child 2`] = `
Object {
  "appendChild": 0,
  "createElement": 3,
  "createElementNS": 0,
  "createTextNode": 0,
  "innerHTML": 0,
  "insertBefore": 3,
  "nodeValue": 0,
  "removeChild": 0,
  "replaceChild": 0,
  "textContent": 1,
}
`;
